app/src/main/java/com/architectcoders/aacboard/ui/fragments/viewmodel/EditBoardCellViewModel.kt

```diff
@@ -0,0 +1,78 @@
+package com.architectcoders.aacboard.ui.fragments.viewmodel
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.architectcoders.aacboard.data.PictogramUI
+import com.architectcoders.aacboard.data.toUIPictogram
+import com.architectcoders.aacboard.domain.data.cell.Cell
+import com.architectcoders.aacboard.domain.data.cell.CellPictogram
+import com.architectcoders.aacboard.domain.use_case.cell.get.GetCellUseCase
+import com.architectcoders.aacboard.domain.use_case.cell.save.SaveCellUseCase
+import com.architectcoders.aacboard.ui.fragments.viewmodel.EditBoardCellViewModel.Destination.*
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+class EditBoardCellViewModel(
+    private val dashboardId: Int,
+    private val row: Int,
+    private val column: Int,
+    private val saveCellUseCase: SaveCellUseCase,
+    private val getCellUseCase: GetCellUseCase,
+) : ViewModel() {
+
+
+    private val _state: MutableStateFlow<EditBoardCellUiState> = MutableStateFlow(
+        EditBoardCellUiState(),
+    )
+    val state: StateFlow<EditBoardCellUiState> get() = _state.asStateFlow()
+
+    init {
+        viewModelScope.launch {
+            val cell: Cell = getCellUseCase(dashboardId, row, column) ?: Cell(row, column, null)
+            updateUiState(_state.value.copy(pictogram = cell.cellPictogram?.toUIPictogram()))
+        }
+    }
+
+    fun onSaveClicked(keyword: String) {
+        _state.value.pictogram?.let {
+            viewModelScope.launch {
+                saveCellUseCase(
+                    dashboardId,
+                    Cell(row, column, CellPictogram(keyword, it.url))
+                )
+                updateUiState(_state.value.copy(destination = BACK))
+            }
+        }
+    }
+
+    fun onPictogramClicked() {
+        updateUiState(_state.value.copy(destination = SEARCH_PICTOGRAM))
+    }
+
+    fun onUpdatePictogram(newPictogram: PictogramUI?) {
+        newPictogram?.let {
+            updateUiState(_state.value.copy(pictogram = newPictogram))
+        }
+    }
+
+    private fun updateUiState(newUiState: EditBoardCellUiState) {
+        _state.update { newUiState }
+    }
+
+    fun resetDestination() {
+        updateUiState(_state.value.copy(destination = null))
```

Comment:
We could use the StateHolder approach here and delegate to the Holder all the view related actions. I was planning to do it on the MainFragment.

Reply:
I need to rewatch Antonio's video to see how to implement StateHolder.

Reply:
Antonio's example is in the `MainFragment` class, he uses the `MainState` as StateHolder for view only events

Reply:
Thanks. New EditBoardCellState class created to clean up the code in fragment class.